contracts/helpers/RebalancerFlashloanSwap.sol

```diff
@@ -0,0 +1,171 @@
+// SPDX-License-Identifier: GPL-3.0
+
+pragma solidity ^0.8.19;
+
+import "./Rebalancer.sol";
+import { IERC3156FlashBorrower } from "oz/interfaces/IERC3156FlashBorrower.sol";
+import { IERC3156FlashLender } from "oz/interfaces/IERC3156FlashLender.sol";
+
+/// @title RebalancerFlashloan
+/// @author Angle Labs, Inc.
+/// @dev Rebalancer contract for a Transmuter with as collaterals a liquid stablecoin and an yield bearing asset
+/// using this liquid stablecoin as an asset
+contract RebalancerFlashloanSwap is Rebalancer, IERC3156FlashBorrower {
+    using SafeERC20 for IERC20;
+    using SafeCast for uint256;
+    bytes32 public constant CALLBACK_SUCCESS = keccak256("ERC3156FlashBorrower.onFlashLoan");
+
+    /// @notice Angle stablecoin flashloan contract
+    IERC3156FlashLender public immutable FLASHLOAN;
+
+    address public swapRouter;
+    address public tokenTransferAddress;
+    uint32 public maxSlippage;
+
+    constructor(
+        IAccessControlManager _accessControlManager,
+        ITransmuter _transmuter,
+        IERC3156FlashLender _flashloan,
+        address _swapRouter,
+        address _tokenTransferAddress,
+        uint32 _maxSlippage
+    ) Rebalancer(_accessControlManager, _transmuter) {
+        if (address(_flashloan) == address(0)) revert ZeroAddress();
+        FLASHLOAN = _flashloan;
+        IERC20(AGTOKEN).safeApprove(address(_flashloan), type(uint256).max);
+
+        swapRouter = _swapRouter;
+        tokenTransferAddress = _tokenTransferAddress;
+        maxSlippage = _maxSlippage;
+    }
+
+    /// @notice Burns `amountStablecoins` for one collateral asset, swap for asset then mints stablecoins from the proceeds of the
+    /// swap
+    /// @dev If `increase` is 1, then the system tries to increase its exposure to the yield bearing asset which means
+    /// burning stablecoin for the liquid asset, swapping for the yield bearing asset, then minting the stablecoin
+    /// @dev This function reverts if the second stablecoin mint gives less than `minAmountOut` of stablecoins
+    /// @dev This function reverts if the swap slippage is higher than `maxSlippage`
+    function adjustYieldExposure(
+        uint256 amountStablecoins,
+        uint8 increase,
+        address collateral,
+        address asset,
+        uint256 minAmountOut,
+        bytes calldata swapCallData
+    ) external {
+        if (!TRANSMUTER.isTrustedSeller(msg.sender)) revert NotTrusted();
+        FLASHLOAN.flashLoan(
+            IERC3156FlashBorrower(address(this)),
+            address(AGTOKEN),
+            amountStablecoins,
+            abi.encode(increase, collateral, asset, minAmountOut, swapCallData)
+        );
+    }
+
+    /**
```

Comment:
It seems we are going to use these functions at multiple places (router, integrators, here). Let's create a genric contract so that we just need to import the implementation everywhere we need without copy pasting.